src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/DefaultRazorComponentSearchEngine.cs

```diff
@@ -0,0 +1,88 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Linq;
+using Microsoft.AspNetCore.Razor.Language;
+using Microsoft.AspNetCore.Razor.Language.Intermediate;
+using Microsoft.CodeAnalysis.Razor;
+using Microsoft.CodeAnalysis.Razor.ProjectSystem;
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer
+{
+    internal class DefaultRazorComponentSearchEngine : RazorComponentSearchEngine
+    {
+        private readonly ProjectSnapshotManager _projectSnapshotManager;
+
+        public DefaultRazorComponentSearchEngine(ProjectSnapshotManager projectSnapshotManager)
+        {
+            _projectSnapshotManager = projectSnapshotManager ?? throw new ArgumentNullException(nameof(projectSnapshotManager));
+        }
+
+        /// <summary>Search for a component in a project based on its tag name and fully qualified name.</summary>
+        /// <remarks>
+        /// This method makes several assumptions about the nature of components. First, it assumes that a component
+        /// a given name `Name` will be located in a file `Name.razor`. Second, it assumes that the namespace the
+        /// component is present in has the same name as the assembly its corresponding tag helper is loaded from.
+        /// Implicitly, this method inherits any assumptions made by TrySplitNamespaceAndType.
+        /// </remarks>
+        /// <param name="tagHelper">A TagHelperDescriptor to find the corresponding Razor component for.</param>
+        /// <param name="documentSnapshot">A document snapshot if found, null otherwise.</param>
+        /// <returns>A tuple containing the document URI and related resources loaded during search.</returns>
+        public override bool TryLocateComponent(TagHelperDescriptor tagHelper, out DocumentSnapshot documentSnapshot)
+        {
+            DefaultRazorTagHelperBinderPhase.ComponentDirectiveVisitor.TrySplitNamespaceAndType(tagHelper.Name, out _, out var typeSpan);
+            var typeName = tagHelper.Name.Substring(typeSpan.Start, typeSpan.Length);
+
+            foreach (var project in _projectSnapshotManager.Projects)
+            {
+                if (!project.FilePath.EndsWith($"{tagHelper.AssemblyName}.csproj", FilePathComparison.Instance))
+                {
+                    continue;
+                }
+
+                foreach (var path in project.DocumentFilePaths)
+                {
+                    // Rule out if not Razor component with correct name
+                    if (!IsPathCandidateForComponent(path, typeName))
+                    {
+                        continue;
+                    }
+
+                    // Get document and code document
+                    documentSnapshot = project.GetDocument(path);
+                    if (!documentSnapshot.TryGetGeneratedOutput(out var razorCodeDocument))
```

Comment:
This is going to return false quite a lot. I'd recommend making the method async and having this be `GetGeneratedOutputAsync`

Reply:
We can't use `ref` parameters when we're async. Thoughts? @NTaylorMullen

Reply:
Also, with this structure, what would the benefit be?